### ISSUE TITLE
The Future of the Curriculum

### ISSUE THREAD

**Issue by KevinMulhern** [TheOdinProject/discussions#1]
We've hit a massive milestone with releasing the Redesigned site yesterday. In doing so we have secured a great platform to build upon for years to come.

Now its time to think about the future of the most important part of TOP, the courses and lessons themselves and decide on the path we are going to take to make them better and ensure that working on them is enjoyable and sustainable.

That in essence is what my goal is with this discussion, how do we make working on the courses sustainable in the ever changing world of web development?

## Our goals with the curriculum
We aim to restructure the curriculum so users have more options. We will be introducing tracks such as:

* Full stack (what we have now)
* Backend development
* Frontend development

On top of this we also want to give our users the option of taking courses individually outside of any track. Each course on its own should take a user from no knowledge of a given topic to at least proficient.

## The problem
We have well over 150 lessons in total, its difficult for a small team to maintain them all. We take care of any pressing issues and have improved some very vital parts of the curriculum but we know theres a lot of lessons and sections that we could improve vastly from user feedback in the chatroom and data about which lessons students don't complete.

On top of this things change fast, technologies, techniques and syntax often go out of date with the introduction and adoption of new ideas.

We need to organize how courses are worked on and by who so we can create the best, most up to date curriculum possible while ensuring that maintainers are not overburdened and have the time and space to focus on one thing with Odin.

## Proposed Solution
Our proposed solution is to have a maintainer or ideally a team of maintainers who's sole responsibility with Odin will be to look after and improve a course of their choosing. This will be similar to how exercism have their language tracks set up.

This structure will ensure that each course is kept up to date and any problems identified are taken care of as well as ensuring that no part of Odin is neglected.

At times, different courses will need to work together for a cohesive and well put together curriculum overall but other than that each maintainer will have complete ownership over their chosen course.

The maintainers will have autonomy and free reign with the vision they have for their course and what lessons that should be in it. From here on out I'll refer to the maintainers of a course as "owners" because thats truly what they will be and we hope that each of them really do feel like they have ownership of their own part of The Odin Project.

Of course if you own something you will be responsible for it... 

### Owner Responsibilities
Odin is open source so no owner should be expected to do it all on their own. She or He, should make use of the community when working on adding new lessons and making and improvements to existing lessons and projects on their course.

A owners main responsibilities will be to:
* Plan out lessons, projects and set out an overall vision for that course. Be the architect of that course essentially.
* Priortize and manage the work that needs done to get the course towards their vision through issues and working with other owners and outside contributors.
* Respond to issues and review pull requests in a timely fashion to ensure outside contributors feel welcome and valued.

## How are we going to do this?
The first steps we should take are splitting our each course from the curriculum repo. This will give owners their own space to work with, will keep issues in a repo focused on one course and ensure owners aren't stepping on each others toes by all working in the one repo.

The initial courses that we will have are as follows:
* Ruby
* Rails
* Javascript
* HTML & CSS
* Git
* Databases
* Electives - for sections and lessons that don't fit in other courses.

There is scope for more courses to be made such as React, Node, Sinatra etc. We will keep the courses focused on our main stack for now but if we are successful with this plan with having owners for each course and if theres interest we will definitely be looking into including more languages that can be used on the backend eg Python, C#, PHP etc. 

### Where will we find owners for the courses
The initial owners will come from the current crop of Odin maintainers, they can choose which ever course interests them and start from there.

But there isn't enough of us to cover all the courses so we will collectively ensure there are issues for people to work on in each course and recognise and value any outside contributors who submit a few quality pull requests to a course by inviting them to the maintainer team and to be the owner of the course they have been helping us build.

If you have any questions, suggestions, concerns, criticisms or anything else to add lets hear it in the comments please 😄

**Comment by ryanford** [comment]
Where in the roadmap is gamification? It would be very helpful if, at the end of each lesson there was a feedback widget or something, where in exchange for some easy point, users could rate or comment on stuff like the relevance, helpfulness, depth/thoroughness, difficulty, how prepared previous lessons made them for the current lesson of the lessons so we could have something quantifiable as to what needs changed.

Also, I'd be happy to help with JS/HTML&CSS

**Comment by KevinMulhern** [comment]
@ryanford-frontend the gamification project, will most likely take place next year. We have some loose ends to tidy up with the redesign and the project submission feature to finish up. When those are done we will be free to focus on the gamification project 💪.

We've talked about a feed back widget before but never thought to combine it with gamification. Thats a excellent idea that will add a lot of value, we should definitely do it 😄 

Glad to have you onboard with the js/html&css courses 🎉

**Comment by mindovermiles262** [comment]
When we add curriculum champions it might be easiest to separate the courses into their own repos

I'd also be happy to help/champion the Ruby section

**Comment by KevinMulhern** [comment]
@mindovermiles262 why do you think it would be easier to separate courses once we have a owner/champion? 😃 

Great to hear you're interested in the Ruby course 💪

**Comment by mindovermiles262** [comment]
Separating the courses would benefit in two ways:
1. Course champions will have an implicit domain.  Many of our courses overlap (git, pair programming, etc) and by setting domains we can ensure that no shared course is missed.  Its easier to keep 100% of a repo updated than it is to keep 25 scattered files updated.

2. Separation will make it easier for users.  When submitting projects I still have a hard time finding the correct file that I'm suppose to edit.  By separating the courses we can minimize this pain point.

I suggest we do this with the change of "ownership" so that we make sure all the assignments are assigned to a repo.

**Comment by KevinMulhern** [comment]
@mindovermiles262 that makes a lot of sense, you've raised some very good points. I didn't even consider the pain point of looking for a repo to submit a solution.

My fear with keeping everything in the one repo while a course doesn't have a owner is that we will be forced to have a lot of different issues by topic in the curriculum repo. 

A compromise would be to split out the big courses, html, js, rails and ruby. And keep the smaller shared courses within the curriculum. Whats your thoughts on that?

**Comment by mindovermiles262** [comment]
Separating big courses sounds good 👍 We should keep the curriculum modular (like any good 
computer program)

**Comment by JonathanYiv** [comment]
I've been brewing about Odin and its future for a little while now, and there's a few topics I want to throw at you all and hear your opinions on.

I know I have not been here too long, but I am passionate about what I believe Odin stands for, an open source curriculum for the people, by the people, and I think these ideas may be of some value.

**TL;DR:** I think the best way to increase the amount of contributors/maintainers of The Odin Project is to grow the community on a massive scale through the use of several mechanisms including: community forums, marketing via online presence, working with nonprofit organizations who need technical skill, offering certification, and accumulating/analyzing/using feedback data. 

Bear with me -- I'm about to do bullet points on why I think each of these are good ideas.

# 1) Our community is small.
This doesn't mean our community isn't great; rather, it's the opposite: I think everyone here is smart, driven, and dedicated.
However, I do think that our community is small ([monthly pulse of theodinproject](https://github.com/TheOdinProject/theodinproject/pulse/monthly)).
This means that progress towards improvement can be slow, because everyone here has their own lives -- we're volunteers and this is open source obviously.
In order to improve that rate of progress, we need more people! If only a select few of the community join the team of contributors/maintainers, that means we need to get **a lot** more people.
So..

# 2) We need more visibility.
I went to a technology conference called [All Things Open](https://allthingsopen.org) this week. When I was networking with people and trying to explain my background, oftentimes they hadn't heard of The Odin Project but they **had** heard of Free Code Camp.

_(Pause: I am not trying to put Odin up to be a competitor to FCC. I believe both have an honorable goal of teaching technical skills to those who might not have the opportunity. While both take a different approach, I think that each has things they can learn from each other.)_

I think it would be awesome if someone reached out to Kevin Mulhern and said "Hey. You run The Odin Project, an open source curriculum that has reached [80,000 people](https://www.theodinproject.com/about)! That's amazing. Would you be interested in coming to talk a little about your project, what you've learned, and the role open source plays in it?" or something.

Then someone goes home, checks it out, recommends it to their nephew who is in college or hates where he is working, and bam: a new community member.
So my thoughts on how we would do that are a multi-pronged approach..

# 3) Working with Nonprofits for Name Recognition
I'll say this upfront: [this is an idea we should 'steal' from FCC](https://www.freecodecamp.org/how-nonprofit-projects-work).
Yes, it has a HUGE amount of oversight and people management required, especially for their specific implementation of the program.
I think a similar program with less resource cost (i.e. effort and time) could be implemented.

This has a huge advantage because it not only offers valuable coding experience/portfolio material to Odinites, but has an outstanding marketing value to it.
"Students from 'The Odin Project' build website to help nonprofit organization SAVE PUPPIES."
Now that's a kind of headline that will draw people to this community.

# 4) Student Retention through Certificates
I understand that ya'll have a poor impression of certificates; they have no real value compared to a portfolio.
I agree!

However, please consider this scenario.
Someone wants to learn how to code.
They do some googling, and they randomly happen on, for example, [SoloLearn](https://play.google.com/store/apps/details?id=com.sololearn&hl=en). 
They also hear about The Odin Project.
From their perspective, both of these things can teach them programming.
(FYI: I was able to earn a 'Ruby certificate' on SoloLearn in about an hour)
However, a quick google search reveals that SoloLearn offers certificates, and you know what -- Odin doesn't.
They think "Hey, maybe an employer will value that certificate."
And so we've just perhaps lost a potential community member who could one day become a contributor.

So, **certificates have value because they make Odin more attractive to prospective students.**
If certificates attract more community members, meaning the community is fuller, then I believe it adds value to Odin.

# 5) Student Retention through Data

I think it is very important to be data-driven.
Let me tell you about some of the statistics we could produce and use!

If a new Odinite signs up, what lesson do they usually drop out on?
Maybe we can email them, and ask them, "Why?"
Then we can improve that lesson. Maybe it's too hard. Maybe it's too boring. We don't know!

What lessons do Odinites spend the most time on?
Are these lessons too hard? Do we need to refactor them? Have the students submitted any pertinent feedback about them?
We could go around refactoring lessons based on what we 'feel' is lacking, but a data-driven approach is more effective.

Those are just some of the most obvious possibilities.
Lastly..

# 6) Community Forum

Chat and forums are very different experiences.
I'm not very often compelled to scroll all the way upward through a few days of missed history.
I may be more compelled to click a topic that says "Vue WILL surpass React and React SUCKS" with 20 posts and 10 posters.
It creates a feeling of longevity to the community, and I think it would be a valuable asset for the community members.

Furthermore, it acts as a form of documentation.
If someone goes to the subforum for the first project: installations, posts about their issues, has it solved by one of you crazy chat guys who know everything, then someone can reference that later on!
In chat, it will disappear into oblivion.

---

I know these are big ideas that would take a huge amount of work.
I'm not highly skilled yet but I plan to be here for a while to pay it forward, and can slowly start building them if everyone is amenable.

If I were to rank them in order of best effect on community growth to work ratio, it might be something like this:
1) Community Forum
2) Certificate
3) Nonprofit Collaboration
4) Data (through Gamification)

Let me know what you think.

**Comment by leosoaivan** [comment]
I'd love to help with Ruby as well. I will be sub-champion #1 :).

**Comment by codyloyd** [comment]
@JonathanYiv a lot of ideas for sure.. and several that we've been discussing for some time 😄 here's my take on them.

**small community/visibility**
I agree with you here.. though I'll admit I am a bit nervous about what would happen if we got too big too quick.  The community is the greatest part of the whole experience (IMO of course!!) and I'm afraid that we could outgrow what makes us special if we grow to quickly.  HOWEVER => more people means more visibility means more people means more contributors which is important.

On that point.. the more contributors we have the harder it is to organize and maintain the quality of contributions which means we'd need more mods/admins and a better structure for collaborating between the mods to make sure we're all on the same page with our goals for the site and curriculum.

to sum up => good ideas here, but we gotta be careful/intentional with how we approach it.

**nonprofits**
another good idea... but something we really need to be careful with.  Don't want to be sending crappy devs to some projects to jack stuff up for them... and who is going to be in charge of organizing such a thing?

**certificates**
we are so torn on this topic lol... like everyone realizes they're pointless, but at the same time if they're going to bring in more students and or help motivate students then it might still be a good idea.  We've been tossing this idea back and forth for as long as I've been involved with Odin...  as far as my opinion: they probably aren't going to _hurt_ anything, so whatever... at the same time if we're going to do them we need to make sure they actually mean something, which means some sort of 'review process'.  Nothing wrong with that, but before we move on we should agree on what the requirements are and how the reviews are going to work.... and we _might_ need more help in the review process if our numbers grow.

**data**
I like data.  We're kind of doing this already with Google Analytics.  We can see how often students mark specific lessons as complete and move on, so we have a pretty good idea of where the big bottlenecks are, but it's hard to know specifically what's causing them to stop where they are.  One of the biggest dropoffs is at the JS/jQuery lesson in WD101, which is why we're working on revamping that right now.

I'd be alllll for coming up with more ways to track this type of stuff for more specific data... Some sort of survey when they haven't logged in in a few days or something maybe?

**Forum**
I like forums and I think having one might be useful.   There's a chance it could harm the chat community we have, but then again I know there are _many_ students who never show up in the chatroom that might benefit. Also, if we think we're going to grow then forums might help us to not kill the vibe in the chatroom by lessening the load there.

So I think I agree, although again there would be a need for more mods... and someone to set the dang thing up in the first place.

**Comment by CouchofTomato** [comment]
Might be worth splitting these topics into their own discussion issues here.

Would it be worthwhile getting a hangout going to try and get everyone on the same page?

**Comment by JonathanYiv** [comment]
What about Discord?
(longer reply incoming later)

**Comment by codyloyd** [comment]
also.. in reply to @KevinMulhern's original post, I'm already tackling the JS side of things.. and it's already in it's own repo here: https://github.com/TheOdinProject/javascript_curriculum

At the moment I'm working on massaging the JS101 material and issues are unfortunately too broad to be actionable... but if anyone wants to jump in please come on over.

On that topic, we should get a clear idea of how the multi-track curriculum is going to work out, especially with regards to sections of each course that are going to be used in multiple tracks (i.e. JS101 and HTML/CSS101)  this isn't a hard problem to solve.. but we should all be on the same page because it could effect how we structure the courses we own.

**Comment by KevinMulhern** [comment]
Thanks for taking the time to write out all your ideas @JonathanYiv, its very clear your passionate about TOP 😄 

As Cody mentioned we've talked about a lot of this within the mod team at one time or another so I'd like to share my thoughts on each point as well.

**Our community is small**
This is something that has improved a lot during my time with Odin, when I started there was no official Odin community. Just a bunch of small unofficial groups on Facebook and Google plus that consisted of students going through the curriculum. I think its important to acknowledge that this is something we've put a lot of time into improving before getting to the crux of it. The reason why we are so small is mainly due to a couple of reasons IMO. One of them is...

**Visibility**
This is a very valid point, we don't have much visibility beyond word of mouth. Thats because we don't do all that much in the way of marketing. The mod team consists mainly of developers, we do what we can but it would be much better if we had someone in the mod team that had a marketers skillset and was interested in upping our marketing game.

The second reason why I believe our visibility is low is due to how damn difficult Odin is which is a good thing in one respect, but a bad thing for on boarding new students. You really need to have grit to make it far in Odin. This is mainly due to the structure and nature of web dev 101's shallow introduction to all topics.

This is something I believe we can tackle with the ownership approach to courses outlined in the top post of this discussion. If we make the curriculum more accessible and ramp up the difficulty level a bit more slowly we should retain more users who will have good things to say about Odin and be more likely to recommend us.

Basically I believe if we have a good product then the growth will follow naturally, we have an ok product with the curriculum right now. But it has flaws, and when we fix those the curriculum will be much better.

**Working wit Non-profits**
This would be a great initiative and an excellent stretch goal for Odin to strive towards. As you mentioned it will take a lot of overhead to get this up and running and to maintain it. Right now it wouldn't be feasible with the size of the team and what priorities we have. But I really would love for us to do something like this someday, we need someone to champion it when the time comes.

**Certificates**
aha we've been talking about this so much lately, my view on this is it won't do any harm and will help students feel safer with choosing TOP over other programs. I agree that we should do it.

**Data**
I've thought __a lot__ about this myself. Currently we have to use google analytics and the rails console to fish out useful data about the bottle necks and problem areas in the curriculum. Which isn't accessible to the whole team and is a pain to do generally.

We've been very focused on student facing features so far, but it's becoming more and more clear that we need two streams of work on the site. One for all the user facing features we build and another for all the things that help the maintainers make a better curriculum. Data falls under this second stream. We plan on making an admin dashboard for the student submission project we have ongoing. This will initially be used to view student submissions that have been "reported" for having an inappropriate link. We can build on top of this to reveal data about lesson completions and as a place to view feedback when we incorporate feed back mechanisms on lessons.

I would also like us to use a/b testing so we can easily and quickly see if changes have a positive or negative effect on metrics.

**Forum**
I agree with this too, I'm sure there are a lot of users who would prefer a forum over the chatroom. The trouble is we will be splitting the community in a sense between the forum and chat room. As Cody mentioned we will need more mods and I believe we will need more active users before this becomes a useful feature to have. Currently we have around 2000 unique users on the site everyday. If we had double this I think a forum feature would work. We could use Discourse for this so building it would be no trouble.

**Comment by JonathanYiv** [comment]
There's a lot to read and think about, but I want to reply for now mostly about the forum, which you may have noticed I listed as #1 for bestest thingy.

I want to make a case for it being the next big feature.

You both mentioned that perhaps the forum might divide the community between the forum and chat room.
From my experience in online communities (mostly gaming admittedly), there are forum users who never would use chat, chat users who would never use the forums, and power users who participate in both.
In that sense, we wouldn't lose chat members because they probably prefer chatting. Instead, we would gain huge value because there is a huge silent majority of the community.

I mentioned this briefly earlier, but it opens up room for so many things.
A) Long-Term Introductions
Right now, if a new person enters the chat, and I'm not there, I'm not going to purposefully engage with that person until they randomly appear later.
They simply disappear, and if they mention having a background in something I am interested in, I'm not going to have the opportunity to reply "Hey, you like ___? Blah blah blah."
Then they get a notification or something and respond back.

B) Documentation
Like I said earlier, common issues with the curriculum will be documented because, well, people like to complain. So they will complain, and we can reference this as a useful form of user feedback to improve.
As beginners, GitHub, issues, and PRs are probably unfamiliar to them, whereas they may have had forum experience and can more easily communicate through it.

C) Suggestions
I'd like to take this repository itself as an example.
We clearly needed a way to discuss important things, scroll back and reference our discussion, think about it, and reply.
So here we are, and I don't think a GitHub repo using issues is the most accessible zone for the normal user.
Chat simply can't offer this functionality either.

D) Discussion
Building on point C, there is simply no means of long-term discussion that is preserved.
Odin has lessons that change infrequently.
A subforum for each lesson and project would enable discussion beyond the chat.
Each lesson has "Discuss this in the Chat!"
If I went into chat right now and said, "I finished APIs tonight and thought this was cool!"
I miiiight get one or two replies.
In a forum, the replies would come slower, but there would be more of them long-term.

<hr>

Those are just a few of the benefits that a forum would bring that the chat cannot.

Lastly, on the topic of moderators. It's certainly a thankless job, but also an invaluable one.

I will volunteer as an initial moderator; I've had experience doing so.

I'm confident that people will naturally rise to the role with time, allowing me to step away later on if I need/want to. People like contributing. People like being important. People like helping.

PS: I also want to help implement the forum if we're gonna do it..

**Comment by CouchofTomato** [comment]
You do all know that gitter has forum rooms right? Just nobody uses them and they aren't well advertised.

**Comment by codyloyd** [comment]
> In that sense, we wouldn't lose chat members because they probably prefer chatting. Instead, we would gain huge value because there is a huge silent majority of the community.

I agree with this.

about the Gitter forums => the visibility of them is miserable. :(  maybe we could find a way to use them more effectively??

AAAAND I've never used discourse at all, but if it's ezpz to set up and use why not give it a shot?

**Comment by codyloyd** [comment]
lets move the forum talk to it's own discussion... I think it's a good topic in general... but we're cluttering up the 'future of the curriculum' discussion here 😄

**Comment by KevinMulhern** [comment]
@JonathanYiv thats a excellent argument, ok I'm convinced we should do it 😄 lets move this forum discussion over to the other discussion. Can you move your comment over there please? it belongs  in there 😄

**Comment by KevinMulhern** [comment]
Back on topic of the main post of this discussion. One of the problems with this ownership approach is we could potentially be creating silos where each course is working on its own things without any considerations for what the curriculum needs as a whole.

One of the big things that attracts students to Odin is the path it provides and the fact everything builds upon each other. We need to be sure we maintain this and don't have courses that don't fit into the path.

The way around this imo, is to have a "Curriculum Goals". For instance one of the curriculum goals we will probably have is to get rid of the web development 101 course (I'll get into the reasons behind this in another discussion) 

To achieve that goal, several different courses will have to do their part to meet that goal. For example The ruby and the rails course will need to create their own installations lesson to replace the one in web dev 101.

Essentially the curriculum goal will provide a direction that we are collectively going in and each course will have its own small part to play to meet that goal. The work they have to do for the curriculum goal should always be the priority for the course but other than that the improvements a owner wants to make will be at their own discretion as long as it doesn't go against the overall curriculum goals.
